### PR TITLE
[MLIR] Add support for Convolution Op without Padding

### PR DESCRIPTION
**Pull request by nmostafa**
Will add padding support later once AffineIfOp builders are supported:
https://github.com/tensorflow/mlir/pull/109

### CODE REVIEW

File: test/cpu_test.cpp [hunk 1]
```diff
@@ -309,7 +309,7 @@ TEST(cpu_test, reshape_layout_optimizations2)
     }
 }
 
-TEST(cpu_test, reshape_layout_optimizations3)
+TEST(cpu_test, MLIR_DISABLE_TEST(reshape_layout_optimizations3))
```
[the remark this answers is not in the document]
Comment (nmostafa) [reply test/cpu_test.cpp]: Those check for specific ngraph optimization that is not applicable with compiled kernel capturing the sub-graph. Nothing we should fix. Just not applicable test.

File: src/contrib/mlir/dialect/ops.cpp [hunk 2]
```diff
@@ -205,6 +205,105 @@ mlir::LogicalResult verifyOp(NGGatherOp* op)
...
+    if (resultRank != imagesRank || resultShape[0] != imagesShape[0] ||
+        resultShape[1] != filtersShape[0])
+    {
+        return op->emitOpError("Invalid result shape");
+    }
+    for (unsigned i = 0; i < resultRank - 2; i++)
+    {
+        unsigned resDim = llvm::divideCeil(padBelowVal[i] + padAboveVal[i] + imagesShape[2 + i] -
+                                               filtersShape[2 + i] + 1,
+                                           stridesVal[i]);
```
[the remark this answers is not in the document]
Comment (nmostafa) [reply src/contrib/mlir/dialect/ops.cpp]: actually applying format caused this. Not sure why

File: test/backend/convolution.in.cpp [hunk 3]
```diff
@@ -66,3 +66,36 @@ NGRAPH_TEST(${BACKEND_NAME}, convolution_outlining)
...
+    auto backend = runtime::Backend::create("${BACKEND_NAME}");
+
+    // Create some tensors for input/output
+    auto a = backend->create_tensor(element::f32, shape_a);
+    copy_data(a, vector<float>{1.0f, 1.0f, 1.0f, 1.0f, 1.0f, 1.0f, 1.0f, 1.0f});
+    auto b = backend->create_tensor(element::f32, shape_b);
+    copy_data(b, vector<float>{1.0f, 1.0f, 1.0f, 1.0f});
+    auto result = backend->create_tensor(element::f32, shape_r);
+
+    vector<float> expected_result{2.0f, 2.0f, 2.0f, 2.0f, 2.0f, 2.0f, 2.0f, 2.0f};
```
[the remark this answers is not in the document]
Comment (nmostafa) [reply test/backend/convolution.in.cpp]: Good idea.

File: src/contrib/mlir/pass/mlir_subgraph_extraction.cpp [hunk 4]
```diff
@@ -369,6 +370,47 @@ bool MLIRSubgraphExtractionPass::is_supported_mlir_op(std::shared_ptr<Node> node
         return true;
     }
 
+    if (TI(ngraph::op::Convolution) == TI(*node))
+    {
+        // No padding for now
+        auto conv_node = static_cast<ngraph::op::Convolution*>(node.get());
+        auto pad_below = conv_node->get_padding_below();
+        for (auto it = pad_below.begin(); it != pad_below.end(); it++)
```
[the remark this answers is not in the document]
Comment (nmostafa) [reply src/contrib/mlir/pass/mlir_subgraph_extraction.cpp]: Changed to `std::all_of`

File: src/contrib/mlir/compiler.cpp [hunk 5]
```diff
@@ -249,21 +250,30 @@ void MLIRCompiler::build_ng_dialect_module()
     dump_mlir_module("nGraph Dialect Construction");
 }
 
-// Converts nGraph shape \p ng_shape to MLIR shape \p mlir_shape.
-static void get_mlir_shape(ngraph::Shape ng_shape, llvm::SmallVectorImpl<int64_t>& mlir_shape)
+template <typename T>
+llvm::SmallVector<int64_t, 4> MLIRCompiler::get_mlir_shape(T ng_shape)
```
[the remark this answers is not in the document]
Comment (nmostafa) [reply src/contrib/mlir/compiler.cpp]: I changed it. This is documented somewhere ?

File: test/cpu_test.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (nmostafa) [reply test/cpu_test.cpp]: This is what fails: `EXPECT_EQ(count_ops_of_type<runtime::cpu::op::ConvertLayout>(cpu_f), 3);`, we are checking for a specific nGraph pass output.

File: src/contrib/mlir/compiler.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (dcaballe) [reply src/contrib/mlir/compiler.cpp]: Unfortunately not that I'm aware of but I've seen this in code reviews and internally. The rationale is that if we enforced a particular SmallVector size in interfaces where the size can be arbitrary, like in this one, we would be missing the point of using a SmallVector data structure. For example, here, if we are sure that the shape is, let's say, 8 in a particular case, the interface should be able to deal with a `SmallVector<int64_t, 8>` efficiently. If a `SmallVector<int64_t, 4>` was enforced by the interface, size 8 would always cause an extra allocation + container copy even though we know for sure the required size.

File: src/contrib/mlir/pass/mlir_subgraph_extraction.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (dcaballe) [reply src/contrib/mlir/pass/mlir_subgraph_extraction.cpp]: It looks great! Thanks!

**Pull request merged**